### PR TITLE
Make pay token more generic and add swapping alerts.

### LINKED ISSUE #60: Add ability to choose pay token

### PR DESCRIPTION
**Pull request by NoahSaso**
Closes #60

### CODE REVIEW

**Comment by NoahSaso** [comment]
# Swap prompt
Screen Shot 2022-02-28 at 2 04 40 PM

## Option to round up to full balance if just swapped and have some extra
Screen Shot 2022-02-28 at 2 05 51 PM

## If their new balance is more than max contribution, cap it there
Screen Shot 2022-02-28 at 2 21 03 PM

## If their their initial contribution was already equivalent to the remaining
Screen Shot 2022-02-28 at 2 09 05 PM

File: ui/components/ContributeForm.tsx [hunk 1]
```diff
@@ -69,81 +88,292 @@ const ContributeFormContents: FunctionComponent<ContributeFormProps> = ({
...
-    1e-6,
-    // fundingTokenPrice is funding tokens (without decimals) per 1 ujuno(x), so invert and divide.
-    // Use ceiling in case 1/fundingTokenPrice is nonzero after the 6th decimal and we need to set a minimum within the 6 decimal range.
-    Math.ceil(1 / (fundingTokenPrice ?? 1)) / 1e6
+  // Minimum contribution is how many non-micro payTokens per micro funding token, since each contribution much return at least 1 micro funding token.
+  const minContribution = convertMicroDenomToDenom(
+    Math.max(
+      // Cannot fund less than 1 micro payToken.
+      1,
+      // fundingTokenPrice is non-micro funding tokens per 1 micro payToken, so invert and convert to non-micro.
```
[the remark this answers is not in the document]
Comment (ezekiiel) [reply ui/components/ContributeForm.tsx:96]: Think in practice this means that the minimum contribution should just be the funding token price. Sending any less than that will cause the contract to round down to zero.

For example, if the token price is 0.5 utokens/ujuno then sending any less than 2 ujuno will result in an error in the contract saying that zero tokens will be returned.

File: ui/components/cards/BalanceRefundJoinCard.tsx [hunk 2]
```diff
@@ -77,10 +78,13 @@ const BalanceRefundJoinCardContents: FunctionComponent<
     watchRefund && watchRefund > 0 && fundingTokenPrice
       ? watchRefund / fundingTokenPrice
       : 0
-  // Minimum refund is how many funding tokens (WITH decimals) per 1 ujuno(x).
-  // fundingTokenPrice is funding tokens (withOUT decimals) per 1 ujuno(x), so divide.
-  // Use ceiling in case fundingTokenPrice is nonzero after the 6th decimal and we need to set a minimum within the 6 decimal range.
-  const minRefund = Math.ceil(fundingTokenPrice ?? 0) / 1e6
+  // Minimum refund is how many non-micro funding tokens per 1 micro payToken.
```
[the remark this answers is not in the document]
Comment (ezekiiel) [reply ui/components/cards/BalanceRefundJoinCard.tsx:81]: Think the logic here is correct though.

File: ui/state/token.ts [hunk 3]
```diff
@@ -81,3 +85,48 @@ export const walletTokenBalance = selectorFamily<
...
+
+export const nativeWalletTokenBalance = selectorFamily<
+  TokenBalanceResponse,
+  string | undefined | null
+>({
+  key: "nativeWalletTokenBalance",
+  get:
+    (tokenDenom) =>
+    async ({ get }) => {
+      if (!tokenDenom) return { balance: null, error: null }
```
Comment (ezekiiel) [inline ui/state/token.ts:97]: What's the point of allowing this to be null if we just pass it through?

**Review by ezekiiel — COMMENTED** [review]
Looking good. Couple comments. Thanks for building this. :)

File: ui/components/ContributeForm.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (NoahSaso) [reply ui/components/ContributeForm.tsx:96]: Yep! Nice catch. I tried translating "with" and "without" decimals to "micro" and "non-micro" and I think I swapped it in these comments. The code reflects the actual math correctly though :))

Not sure why I'm clamping this lower value at 1... Math.ceil should take care of that already. Will fix.

File: ui/components/cards/BalanceRefundJoinCard.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (NoahSaso) [reply ui/components/cards/BalanceRefundJoinCard.tsx:81]: Yeah the comments are actually correct this time (compared to the contribution comments which are wrong). The refund entry field is in terms of denom, not micro denom, so min refund is non-micro funding tokens per 1 micro pay token.

File: ui/services/chain.ts [hunk 4]
```diff
@@ -166,3 +172,68 @@ export const createDAOProposalForCampaign = async (
...
   return findAttribute(response.logs, "wasm", "proposal_id").value
 }
+
+export const getNativeTokenBalance = async (
+  client: SigningCosmWasmClient,
+  walletAddress: string,
+  token: PayToken
+): Promise<number> => {
+  const coin = await client.getBalance(walletAddress, token.denom)
+  return Number(coin?.amount ?? 0) / Math.pow(10, token.decimals)
```
[the remark this answers is not in the document]
Comment (NoahSaso) [reply ui/services/chain.ts:182]: Funny, this was literally the code that made me want to make those helper functions... great catch.

File: ui/state/token.ts (hunk 3 above)
Comment (NoahSaso) [reply ui/state/token.ts:97]: Hooks can't be conditionally called in components, so the alternative would be passing an empty string to the selector and running more code (the few lines below this that try to find a matching payToken) before just passing null through anyways.

File: ui/services/chain.ts [hunk 5]
```diff
@@ -166,3 +172,68 @@ export const createDAOProposalForCampaign = async (
...
+    swapPrice,
+    baseToken.decimals
+  )
+
+  const microInputAmount = Math.round(
+    inputAmount * Math.pow(10, baseToken.decimals)
+  ).toString()
+  const microMinOutputAmount = Math.floor(
+    minOutput * Math.pow(10, outputToken.decimals)
+  ).toString()
```
[the remark this answers is not in the document]
Comment (NoahSaso) [reply ui/services/chain.ts:221]: I believe `convertDenomToMicroDenom` but yes.

**Pull request merged**